server/protocols/jmap/src/main/java/org/apache/james/jmap/event/PropagateLookupRightListener.java

```diff
@@ -0,0 +1,130 @@
+        this.storeRightManager = storeRightManager;
+        this.mapperFactory = mapperFactory;
+    }
+
+    @Override
+    public ListenerType getType() {
+        return ListenerType.ONCE;
+    }
+
+    @Override
+    public ExecutionMode getExecutionMode() {
+        return ExecutionMode.ASYNCHRONOUS;
+    }
+
+    @Override
+    public void event(Event event) {
+        MailboxSession mailboxSession = event.getSession();
+
+        if (event instanceof MailboxACLUpdated) {
+            MailboxACLUpdated aclUpdateEvent = (MailboxACLUpdated) event;
+
+            updateLookupRightOnParent(mailboxSession, event.getMailboxPath(), aclUpdateEvent.getAclDiff());
+        } else if (event instanceof MailboxRenamed) {
+            MailboxRenamedEventImpl renamedEvent = (MailboxRenamedEventImpl) event;
+
+            updateLookupRightOnParent(mailboxSession, renamedEvent.getNewPath(), renamedEvent.getMailbox().getACL());
+        }
+    }
+
+    private void updateLookupRightOnParent(MailboxSession mailboxSession, MailboxPath mailboxPath, MailboxACL mailboxACL) {
+        getParent(mailboxPath, mailboxSession)
+            .forEach(parentMailboxPath -> updateLookupRight(parentMailboxPath, mailboxSession,
+                mailboxACL.getEntries()
+                    .entrySet()
+                    .stream()
+                    .map(entry -> new MailboxACL.Entry(entry.getKey(), entry.getValue()))));
+    }
+
+    private void updateLookupRightOnParent(MailboxSession mailboxSession, MailboxPath mailboxPath, ACLDiff aclDiff) {
+        getParent(mailboxPath, mailboxSession)
+            .forEach(parentMailboxPath -> {
+                updateLookupRight(parentMailboxPath, mailboxSession, aclDiff.addedEntries());
+                updateLookupRight(parentMailboxPath, mailboxSession, aclDiff.changedEntries());
+            });
+    }
+
+    private void updateLookupRight(MailboxPath mailboxPath, MailboxSession session, Stream<MailboxACL.Entry> entries) {
+        entries
+            .filter(entry -> !entry.getKey().isNegative())
+            .filter(entry -> entry.getValue().contains(Right.Lookup))
+            .forEach(entry -> updateRight(mailboxPath, session, entry));
+    }
+
+    private Stream<MailboxPath> getParent(MailboxPath mailboxPath, MailboxSession mailboxSession) {
+        return mailboxPath.getHierarchyLevels(mailboxSession.getPathDelimiter())
+            .stream()
+            .filter(hierarchyMailboxPath -> !hierarchyMailboxPath.equals(mailboxPath));
+    }
+
+    private void updateRight(MailboxPath mailboxPath, MailboxSession session, Entry entry) {
+        try {
+            boolean hasLookupRight = mapperFactory.getMailboxMapper(session)
+                .findMailboxByPath(mailboxPath)
+                .getACL()
+                .getEntries()
+                .getOrDefault(entry, MailboxACL.NO_RIGHTS)
+                .contains(Right.Lookup);
+
+            if (!hasLookupRight) {
```

Comment:
why do we care about previous state ?

Reply:
We will not update mailbox's right if it had already 'l' lookup, is that correct behavior?

Reply:
putting a `l` right on top of another one probably would not hurt, right ? any other opinion ?

Reply:
changed